### PR TITLE
Speaker diarization with whisper transcription

### PR DESCRIPTION
**Pull request by Jiltseb**
This PR addresses the feature request https://github.com/mobiusml/aana_sdk/issues/176.

1. Combines output from Speaker Diarization Deployment and Whisper deployment.
2. The post-processing algorithm generates diarized transcription as shown in the example notebook.
3. Added test for the post-processing algorithm.
4. Changes in docs

### CODE REVIEW

File: aana/processors/speaker.py [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+        word_count = 0  # Reset word count
+
+    else:
+        # Append word to the current segment
+        current_segment.time_interval.end = word_info.time_interval.end
+        current_segment.text += f"{word_info.word}"
+        current_segment.words.append(word_info)
+        word_count += 1
+
+    # If sentence-ending punctuation is found, finalize the segment
```
Comment (movchan74) [inline aana/processors/speaker.py:397]: Why is commented out? Can we remove it?

File: aana/processors/speaker.py [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+    Returns:
+        word_speaker_mapping: (list[AsrWord]): Realigned word-speaker mappings.
+    """
+    new_segments = [segment.words for segment in transcription["segments"]]
+    word_speaker_mapping = [word for segment in new_segments for word in segment]
+    words_list = [item.word for item in word_speaker_mapping]
+    speaker_list = [item.speaker for item in word_speaker_mapping]
+
+    # Fill missing speaker labels by finding the nearest speaker
+    for i, item in enumerate(word_speaker_mapping):
```
Comment (movchan74) [inline aana/processors/speaker.py:217]: Is it the same as fill_nearest in get_speaker_for_interval?

File: aana/processors/speaker.py [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+        # Assign speaker to segment
+        segment.speaker = get_speaker_for_interval(
+            diarized_output["segments"],
+            segment.time_interval.start,
+            segment.time_interval.end,
+            fill_nearest,
+        )
+
+        # Assign speakers to words within the segment
+        if segment.words:
```
Comment (movchan74) [inline aana/processors/speaker.py:70]: I saw `whisper_params.word_timestamps = True` in the notebook. As I understand none of this will work without word level predictions. Then we should add check somewhere to make sure that word-level predictions are included. Maybe we can check segment text and if it's not empty but words are empty list then we throw exception.

**Review by movchan74 — CHANGES_REQUESTED** [review]

File: aana/deployments/pyannote_speaker_diarization_deployment.py [hunk 4]
```diff
@@ -69,9 +71,21 @@ async def apply_config(self, config: dict[str, Any]):
...
-        # load model using pyannote Pipeline
-        self.diarize_model = Pipeline.from_pretrained(self.model_id)
-        self.diarize_model.to(torch.device(self.device))
+        try:
+            # load model using pyannote Pipeline
+            self.diarize_model = Pipeline.from_pretrained(self.model_id)
+
+            # Check if model, log error if None.
+            if self.diarize_model is None:
+                logging.error(
```
[the remark this answers is not in the document]
Comment (Jiltseb) [reply aana/deployments/pyannote_speaker_diarization_deployment.py:80]: What about:
`        try:
            # load model using pyannote Pipeline
            self.diarize_model = Pipeline.from_pretrained(self.model_id)

            if self.diarize_model:
                self.diarize_model.to(torch.device(self.device))

        except BaseException as e:
            e.add_extra({self.model_id: "Failed to load the model."})
            raise e`

[the remark this answers is not in the document]
Comment (movchan74) [reply aana/deployments/pyannote_speaker_diarization_deployment.py:80]: Just create a new exception. 
And there seems to be an exception GatedRepoError in the transformers. Why isn't it raised? Then we can just use that.

File: aana/processors/speaker.py [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+        key=lambda data: data[
+            "word_count"
+        ],  # Change this to 'contribution_time' if needed
+    )
+
+    major_segment = major_segment_data["segment"]
+    return major_segment.confidence, major_segment.no_speech_confidence
+
+
+def merge_consecutive_speaker_segments(
```
[the remark this answers is not in the document]
Comment (Jiltseb) [reply aana/processors/speaker.py:475]: `combine_homogeneous_speaker_segments` function is used to merge speaker diarization segments and account for possible overlapping. The function `merge_consecutive_speaker_segments` merges consecutive ASR segments that has a single speaker. A better name would be `combine_homogeneous_speaker_diarization_segments``and `combine_homeogeneous_speaker_asr_segments` respectively.

File: aana/processors/speaker.py (hunk 2 above)
Comment (Jiltseb) [reply aana/processors/speaker.py:217]: No, both are different. This one finds the speaker for the next word(s) if the current word does not have a speaker assigned to it. `get_speaker_for_interval` assigns each word/segment a corresponding speaker if there is max overlap between the word and the diarization segment. If fill_nearest is True, it assigns each word a corresponding speaker even if there is zero overlap.

Basically, in the first case, `the near to a word` is concerning neighboring words. In the latter,` near to a word/segment` means the overlap of diarization segment with the word, and if it doesn't exist, look for the nearest diarization segment (has nothing to do with a nearby word).

File: aana/processors/speaker.py [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+    AsrTranscriptionInfo,
+    AsrWord,
+)
+from aana.core.models.speaker import SpeakerDiarizationSegment
+from aana.core.models.time import TimeInterval
+
+# Utility functions for speaker-related processing in audio
+
+
+# Redefine SpeakerDiarizationOutput and WhisperOutput to prevent circular imports
```
[the remark this answers is not in the document]
Comment (Jiltseb) [reply aana/processors/speaker.py:16]: I wanted to use type annotations for the functions. Or is it not needed for processing functions?

**Comment by Jiltseb** [comment]
I have changed the post-processing function into a class to keep the logic organized and the module-specific functions private. However, it has not been moved to diarization deployment because it will kill the flexibility to run the diarization and whisper models in parallel if needed. I think we can keep it this way until we have a specific SDK component that can combine the deployments and additional functions.

File: aana/processors/speaker.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (movchan74) [reply aana/processors/speaker.py]: If you only use imports for type checking then you can use `typing.TYPE_CHECKING`. See https://github.com/mobiusml/aana_sdk/blob/8cb6ac0c4aaef05e70ec16b8ec575985ba7c8a49/aana/exceptions/io.py#L5

File: aana/processors/speaker.py [hunk 7]
```diff
@@ -0,0 +1,621 @@
...
+        self.diarized_segments = diarized_segments
+        self.transcription_segments = transcription_segments
+        self.merge = merge
+
+        # Check if inputs are valid:
+        for segment in self.transcription_segments:
+            if segment.text and not segment.words:
+                raise ValueError("Word-level timestamps are required for diarized ASR.")  # noqa: TRY003
+
+    def process(self) -> list[AsrSegment]:
```
Comment (movchan74) [inline aana/processors/speaker.py:48]: Consider using `classmethod` if the class primarily serves as a utility for post-processing and doesn't need to maintain state. Here are some considerations for making this change:

### Advantages of Using `classmethod`:

1. **Simpler Usage:**
   - If your class doesn't need to maintain state across multiple method calls and is only used to perform a single processing task, using a `classmethod` would make the API simpler.
   - Users can call `ASRPostProcessingForDiarization.process(diarized_segments, transcription_segments, merge=False)` directly without having to create an instance.

2. **No State Management:**
   - If the attributes `diarized_segments`, `transcription_segments`, and `merge` are only needed within the scope of the `process` method, using a `classmethod` avoids the unnecessary creation and management of object state.

3. **Utility-Class Style:**
   - For classes that act more like utility functions and don't hold persistent data or configuration, `classmethod` is a better fit. This makes it clear that the class is used for its functionality rather than as an object with attributes.

### How It Would Look:

Here's how you could refactor your class:

```python
class ASRPostProcessingForDiarization:
    """Class to handle post-processing for diarized ASR output by combining diarization and transcription segments.

    The post-processing involves assigning speaker labels to transcription segments and words, aligning speakers
    with punctuation, optionally merging homogeneous speaker segments, and reassigning confidence information to the segments.
    """

    @classmethod
    def process(
        cls,
        diarized_segments: list[SpeakerDiarizationSegment],
        transcription_segments: list[AsrSegment],
        merge: bool = False,
    ) -> list[AsrSegment]:
        """Executes the post-processing pipeline that combines diarization and transcription segments.

        This method performs the following steps:
        1. Assign speaker labels to each segment and word in the transcription based on the diarization output.
        2. Align speakers with punctuation.
        3. Create new transcription segments by combining the speaker-labeled words.
        4. Optionally, merge consecutive speaker segments.
        5. Add confidence and no_speech_confidence to the new segments.

        Args:
            diarized_segments (list[SpeakerDiarizationSegment]): Contains speaker diarization segments.
            transcription_segments (list[AsrSegment]): Transcription segments.
            merge (bool): If True, merges consecutive speaker segments in the final output. Defaults to False.

        Returns:
            list[AsrSegment]: Updated transcription segments with speaker information per segment and word.
        """
        # intro: validation checks
        if not transcription_segments or not diarized_segments:
            return transcription_segments

        # 1. Assign speaker labels to each segment and word
        speaker_labelled_transcription = cls._assign_word_speakers(
            diarized_segments, transcription_segments
        )

        # 2. Align speakers with punctuation
        word_speaker_mapping = cls._align_with_punctuation(
            speaker_labelled_transcription
        )

        # 3. Create new transcription segments with speaker information
        segments = cls._create_speaker_segments(word_speaker_mapping)

        # 4. Assign confidence variables to the new segments
        segments = cls._add_segment_variables(segments, transcription_segments)

        # Optional: Merge consecutive speaker segments
        if merge:
            segments = cls._combine_homogeneous_speaker_asr_segments(segments)

        return segments

    @staticmethod
    def _assign_word_speakers(diarized_segments, transcription_segments):
        # Method logic here
        pass

    @staticmethod
    def _align_with_punctuation(speaker_labelled_transcription):
        # Method logic here
        pass

    @staticmethod
    def _create_speaker_segments(word_speaker_mapping):
        # Method logic here
        pass

    @staticmethod
    def _add_segment_variables(segments, transcription_segments):
        # Method logic here
        pass

    @staticmethod
    def _combine_homogeneous_speaker_asr_segments(segments):
        # Method logic here
        pass
```

### Considerations Before Changing:

1. **Initialization Logic:**
   - If you need to perform any initialization or validation logic that requires setting up the class, then a `classmethod` may not be suitable. You would lose the ability to store state or configuration that might be necessary for the processing.

2. **Extensibility:**
   - If you anticipate needing to extend this class with additional attributes or state, keeping an `__init__` constructor might be more future-proof.

3. **Usability:**
   - Depending on the complexity and number of parameters, a `classmethod` might be less readable if it involves many arguments. If the `process` method requires many parameters, it might make more sense to use a constructor to set up the instance once.

In summary, using a `classmethod` is a good idea if you want a simpler, stateless utility-class style usage. If the class might need to manage state or if you foresee expanding it with more features, you might want to keep the constructor.

File: docs/pages/model_hub/asr.md [hunk 8]
```diff
@@ -49,4 +49,68 @@ Here are some other possible configurations for the Whisper deployment:
...
+
+### Diarized ASR
+
+Diarized transcription can be generated by using [WhisperDeployment](./../../reference/deployments.md#aana.deployments.WhisperDeployment) and [PyannoteSpeakerDiarizationDeployment](./../../reference/deployments.md#aana.deployments.PyannoteSpeakerDiarizationDeployment) and combining the timelines using post processing with [ASRPostProcessingForDiarization](./../../reference/processors.md).
+
+Example configuration for the PyannoteSpeakerDiarization model is available at [Speaker Diarization](./speaker_recognition.md) model hub.
+
+You can simply define the model deployments and the endpoint to transcribe the video with diarization. Below code snippet shows the custom endpoint class `TranscribeVideoWithDiarEndpoint` to combine the outputs from ASR and diarization deployments:
+
+    ```python
```
Comment (movchan74) [inline docs/pages/model_hub/asr.md:62]: The code is not displayed properly in the docs, probably because of the extra indentation. Please start docs with `poetry run mkdocs serve` to see if it displays properly.

Comment (movchan74) [reply docs/pages/model_hub/asr.md:62]: Also can we boil it down to the basics, just to show how to produce and combine ASR and SD segments, and refer to notebook for detailed example?

File: docs/pages/model_hub/asr.md [hunk 9]
```diff
@@ -49,4 +49,68 @@ Here are some other possible configurations for the Whisper deployment:
...
+        )
+        updated_segments = post_processor.process()
+        output_segments = [
+            s.model_dump(include=["text", "time_interval", "speaker"])
+            for s in updated_segments
+        ]
+
+        return {"segments": output_segments}
+    ```
+An example notebook on diarized transcription is available at `notebooks/diarized_transcription_example.ipynb`.
```
Comment (movchan74) [inline docs/pages/model_hub/asr.md:115]: Add a link to the notebook.

File: docs/pages/model_hub/asr.md [hunk 10]
```diff
@@ -49,4 +49,68 @@ Here are some other possible configurations for the Whisper deployment:
...
     )
-    ```
\ No newline at end of file
+    ```
+
+### Diarized ASR
+
+Diarized transcription can be generated by using [WhisperDeployment](./../../reference/deployments.md#aana.deployments.WhisperDeployment) and [PyannoteSpeakerDiarizationDeployment](./../../reference/deployments.md#aana.deployments.PyannoteSpeakerDiarizationDeployment) and combining the timelines using post processing with [ASRPostProcessingForDiarization](./../../reference/processors.md).
+
+Example configuration for the PyannoteSpeakerDiarization model is available at [Speaker Diarization](./speaker_recognition.md) model hub.
```
Comment (movchan74) [inline docs/pages/model_hub/asr.md:58]: Point to diarization section like `speaker_recognition/#speaker-diarization-sd-models`.

File: docs/pages/model_hub/speaker_recognition.md [hunk 11]
```diff
@@ -35,7 +35,7 @@ The PyAnnote speaker diarization models are gated, requiring special access. To
     To get your Hugging Face access token, visit the [Hugging Face Settings - Tokens](https://huggingface.co/settings/tokens).
 
 
-## Example Configurations
+### Example Configurations
```
Comment (movchan74) [inline docs/pages/model_hub/speaker_recognition.md:38]: Why did you change the level?

File: docs/pages/model_hub/speaker_recognition.md [hunk 12]
```diff
@@ -53,4 +53,9 @@ As an example, let's see how to configure the Pyannote Speaker Diarization deplo
             sample_rate=16000,
         ).model_dump(mode="json"),
     )
-    ```
\ No newline at end of file
+    ```
+
+### Diarized ASR
```
Comment (movchan74) [inline docs/pages/model_hub/speaker_recognition.md:58]: This should probably be at second level.

File: docs/pages/model_hub/speaker_recognition.md [hunk 13]
```diff
@@ -53,4 +53,9 @@ As an example, let's see how to configure the Pyannote Speaker Diarization deplo
...
+
+Speaker Diarization output can be combined with ASR to generate transcription with speaker information. Further details and code snippet are available in [ASR model hub](./asr.md).
+
```
Comment (movchan74) [inline docs/pages/model_hub/speaker_recognition.md:61]: Link it to the exact section: `./asr.md/#diarized-asr`.

File: aana/processors/speaker.py [hunk 14]
```diff
@@ -0,0 +1,621 @@
...
+    AsrWord,
+)
+from aana.core.models.speaker import SpeakerDiarizationSegment
+from aana.core.models.time import TimeInterval
+
+# Define sentence ending punctuations to split segments at sentence endings:
+sentence_ending_punctuations = ".?!"
+
+
+class ASRPostProcessingForDiarization:
```
Comment (movchan74) [inline aana/processors/speaker.py:14]: Just a small stylistic nitpick but we usually capitalise only first letter like `AsrPostProcessingForDiarization` instead of `ASRPostProcessingForDiarization`.

File: aana/processors/speaker.py [hunk 15]
```diff
@@ -0,0 +1,621 @@
...
+                current_segment.speaker, f"SPEAKER_{speaker_counter:02d}"
+            )
+            for word in current_segment.words:
+                word.speaker = current_segment.speaker
+            merged_segments.append(current_segment)
+
+        return merged_segments
+
+
+# speaker diarization model occationally produce overlapping chunks/ same speaker segments,
```
Comment (movchan74) [inline aana/processors/speaker.py:561]: What does this comment refers to? combine_homogeneous_speaker_diarization_segments?

File: aana/processors/speaker.py [hunk 16]
```diff
@@ -0,0 +1,621 @@
...
+            merge (bool): If True, merges consecutive speaker segments in the final output. Defaults to False.
+        """
+        self.diarized_segments = diarized_segments
+        self.transcription_segments = transcription_segments
+        self.merge = merge
+
+        # Check if inputs are valid:
+        for segment in self.transcription_segments:
+            if segment.text and not segment.words:
+                raise ValueError("Word-level timestamps are required for diarized ASR.")  # noqa: TRY003
```
Comment (movchan74) [inline aana/processors/speaker.py:46]: Can we add test for that?

File: aana/processors/speaker.py (hunk 15 above)
Comment (Jiltseb) [reply aana/processors/speaker.py:561]: Yes

File: aana/processors/speaker.py (hunk 14 above)
Comment (Jiltseb) [reply aana/processors/speaker.py:14]: renamed to `PostProcessingForDiarizedAsr` as it is more meaningful.

File: docs/pages/model_hub/speaker_recognition.md (hunk 13 above)
Comment (Jiltseb) [reply docs/pages/model_hub/speaker_recognition.md:61]: Done

File: docs/pages/model_hub/asr.md (hunk 10 above)
Comment (Jiltseb) [reply docs/pages/model_hub/asr.md:58]: Done

File: docs/pages/model_hub/asr.md (hunk 9 above)
Comment (Jiltseb) [reply docs/pages/model_hub/asr.md:115]: Done, it will show once merged to main as the link is: `https://github.com/mobiusml/aana_sdk/tree/main/notebooks/diarized_transcription_example.ipynb`

File: docs/pages/model_hub/asr.md (hunk 8 above)
Comment (Jiltseb) [reply docs/pages/model_hub/asr.md:62]: Removed the class and just retained the combination logic

File: aana/processors/speaker.py (hunk 7 above)
Comment (Jiltseb) [reply aana/processors/speaker.py:48]: Makes sense to add it as a `classmethod` at this point.

File: aana/processors/speaker.py (hunk 15 above)
Comment (movchan74) [reply aana/processors/speaker.py:566]: Can we move it to the docstring? Also, there are some typos in the text.

File: docs/pages/model_hub/asr.md [hunk 17]
```diff
@@ -49,4 +49,37 @@ Here are some other possible configurations for the Whisper deployment:
...
+
+You can simply define the model deployments and the endpoint to transcribe the video with diarization. Below code snippet shows the how to combine the outputs from ASR and diarization deployments:
+
+
+```python
+from aana.processors.speaker import PostProcessingForDiarizedAsr
+
+# diarized transcript requires word_timestamps from ASR
+whisper_params.word_timestamps = True
+transcription = await self.asr_handle.transcribe(
```
Comment (movchan74) [inline docs/pages/model_hub/asr.md:68]: Can we add some comment about where asr_handle and diar_handle come from? Just to make sure that the deployments that you are mentioned there.

File: docs/pages/model_hub/asr.md [hunk 18]
```diff
@@ -49,4 +49,37 @@ Here are some other possible configurations for the Whisper deployment:
...
+    audio=audio, params=whisper_params
+)
+diarized_output = await self.diar_handle.diarize(
+    audio=audio, params=diar_params
+)
+updated_segments = PostProcessingForDiarizedAsr(
+    diarized_segments=diarized_output["segments"],
+    transcription_segments=transcription["segments"],
+)
+output_segments = [
```
Comment (movchan74) [inline docs/pages/model_hub/asr.md:78]: Not sure if we need that filtering here. Instead I would add example output as a comment.

File: aana/processors/speaker.py (hunk 15 above)
Comment (movchan74) [reply aana/processors/speaker.py:566]: Can we move it to the docstring? Also, there are some typos in the text.

File: docs/pages/model_hub/asr.md (hunk 17 above)
Comment (Jiltseb) [reply docs/pages/model_hub/asr.md:68]: This is why I added the full Endpoint Class initially. It is nothing but initializing the handles and running them.

Comment (movchan74) [reply docs/pages/model_hub/asr.md:71]: Maybe like this:

```python
# asr_handle is an AanaDeploymentHandle for WhisperDeployment
transcription = await self.asr_handle.transcribe(
    audio=audio, params=whisper_params
)
# diar_params is an AanaDeploymentHandle for PyannoteSpeakerDiarizationDeployment
diarized_output = await self.diar_handle.diarize(
    audio=audio, params=diar_params
)
```

File: docs/pages/model_hub/asr.md [hunk 19]
```diff
@@ -49,4 +49,76 @@ Here are some other possible configurations for the Whisper deployment:
...
+)
+diarized_output = await self.diar_handle.diarize(
+    audio=audio, params=diar_params
+)
+updated_segments = PostProcessingForDiarizedAsr.process(
+    diarized_segments=diarized_output["segments"],
+    transcription_segments=transcription["segments"],
+)
+
+# updated_segments will have speaker information as well:
```
[the remark this answers is not in the document]
Comment (Jiltseb) [reply docs/pages/model_hub/asr.md:82]: yeah I will add: `from aana.core.models.base import pydantic_to_dict
`

[the remark this answers is not in the document]
Comment (movchan74) [reply docs/pages/model_hub/asr.md]: Still it wouldn't produce the output you added and even adds more complexity. Just check what updated_segments actually returns and copy it there (not in full obviously).

File: aana/processors/speaker.py [hunk 20]
```diff
@@ -0,0 +1,633 @@
...
+        segment.speaker = get_speaker_for_interval(
+            diarized_output["segments"],
+            segment.time_interval.start,
+            segment.time_interval.end,
+            fill_nearest,
+        )
+
+        # Assign speakers to words within the segment
+        if segment.words:
+            for word in segment.words:
```
[the remark this answers is not in the document]
Comment (Jiltseb) [reply aana/processors/speaker.py]: Yes, every word needs to be aligned with a corresponding speaker segment. Not high compared to the diarization time.

**Review by movchan74 — APPROVED** [review]
Looks good! Great work 👍
